### ISSUE TITLE
Two column vs. one column

### ISSUE THREAD

**Issue by jonsaadfalcon** [kermitt2/grobid#419]
Currently, it is very difficult to accurately parse a document that has two columns of text since grobid analyzes it as one column. Has anyone found a way to get around it?

**Comment by kermitt2** [comment]
GROBID supports normally quite well 2-columns documents, reading order is correctly reconstructs in most of the cases for scholar articles in benchmarks against PMC or some publisher native xml. Do you have some particular examples?

**Comment by jonsaadfalcon** [comment]
Sure, I attached a scholarly article to this comment. Please test it and tell me your results
[Parkinson's Article.pdf](https://github.com/kermitt2/grobid/files/3050945/Parkinson.s.Article.pdf)

**Comment by kermitt2** [comment]
Thanks! Indeed it doesn't work at all for this document, I will have a better look - it might work in a close future as we are using [pdfalto](https://github.com/kermitt2/pdfalto) now for low level parsing of pdf, which has better a reading order reconstruct normally.

**Comment by jonsaadfalcon** [comment]
Awesome when should the next update be available, if you have some approximate estimate?

**Comment by OgnjenMilicevic** [comment]
I wanted to add another example of the same problem.
[Increased oxidative stress in mouse kidneys with unilateral ureteral obstruction.pdf](https://github.com/kermitt2/grobid/files/3464931/Increased.oxidative.stress.in.mouse.kidneys.with.unilateral.ureteral.obstruction.pdf)

**Comment by Vitaliy-1** [comment]
Hi @kermitt2,

I also can reproduce the problem on some of PDFs, although I'd say that the number of problematic articles isn't big. Sometimes PDFAlto completely fails to recognize 2 column text correctly, sometimes only part of the text but it happens. Let me know if you'd need some help on debugging.

**Comment by Vitaliy-1** [comment]
As I understand, this method overrides the one from `xpdf`:https://github.com/kermitt2/pdfalto/blob/master/src/XmlAltoOutputDev.cc#L8290

And the problem appears to be even before it's called. As `xpdf` output looks good, I'm, making an assumption that the fix, some sort of reordering, is done in `xpdf` somewhere after this call: https://github.com/kermitt2/xpdf-4.00/blob/master/xpdf/Gfx.cc#L3842 but I haven't found it yet.

**Comment by Vitaliy-1** [comment]
@kermitt2, the split in `xpdf` seems to be done in this iteration: https://github.com/kermitt2/xpdf-4.00/blob/master/xpdf/TextOutputDev.cc#L2861-L2870. As you can see there is a block of code for horizontal spliting, is it possible to implement something similar for PDFAlto?

**Comment by Vitaliy-1** [comment]
During the annotation of our set, I've noticed problems with near 20% of all 2-column articles. most common of them:
1) 2 columns are parsed as one, line by line.
2) Incorrect ordering of paragraphs, e.g., first paragraphs from the left column, then the first abstract from the second, etc. I believe the distance between abstracts has some meaning for the xpdf.
3) Misplaced columns - right is parsed first. 

Using the xpdf's method of sorting, which was implemented but [commented out](https://github.com/kermitt2/pdfalto/blob/master/src/XmlAltoOutputDev.cc#L7740-L7742) on the master branch, drastically reduces the number of errors; I would say it's 10 times more accurate. The one drawback is that sometimes it places redundant spaces before symbols, like brackets. It leads to slight changes in PDFAlto output - 2 tokens instead of one.

**Comment by kermitt2** [comment]
Thank you Vitaliy! 

Yes it's problems relevant to pdfalto, just some remarks:

1. the recognition of columns could be improved a lot in pdfalto I think (it's pretty basic right now), for instance checking that lines are aligned and refining more dynamically the thresholds

2. in pdfalto (it's not related to xpdf I think), there are some heuristics at block level to be sure that we catch two columns for different areas of the documents. Its not always easy to decide when the second column should be considered in the flow of the documents (it's not always at the top of the page). 

3. Probably a question of stream order vs reading order?

The `dumpInReadingOrder` in pdfalto creates other issues if I remember well (when you write xpdf, it's pdfalto?) and it is not as complete as the default `dump` (for line detection in particular, rotations, ...), and as an alternative I wrote a reading order for ordering only the blocks (`addBlockInReadingOrder`). Have you tried maybe to compare the two? The `dumpInReadingOrder` method would need to be completed.

**Comment by Vitaliy-1** [comment]
Yeah, I've tried comparing them but the result was the same for those issues. But the sample wasn't big. The real difference was when I added xpdf sorting algorithm, which is done on chars with sime kind of tree sort. As I remember it's implemented here: https://github.com/kermitt2/pdfalto/blob/master/src/XmlAltoOutputDev.cc#L3405 and here https://github.com/kermitt2/pdfalto/blob/master/src/XmlAltoOutputDev.cc#L3555. `dumpInReadingOrder` uses it here: https://github.com/kermitt2/pdfalto/blob/master/src/XmlAltoOutputDev.cc#L4673 but dump method doesn't and uses own way to form words without such a sorting. It was initially bypassed in the pdf2xml code. I'm not sure why they omitted this part, probably because if issues that you've encountered. 

When I'm writing xpdf, I'm referring to its methods, which are used by pdfalto.

**Comment by Vitaliy-1** [comment]
In short the `dumpInReadingOrder` bypasses the character sequence through all sorting methods that `xpdf` uses but the `dump` method doesn't.

**Comment by kindofausername** [comment]
Is there any progress on this issue? I'm using the latest Docker version, so there have been no code changes on my side. I'm encountering incorrect ordering of paragraphs in some PDFs. Is there a way to exclusively use the Deep Learning implementation for body extraction?

Since this thread is quite old, is it now possible to use the Deep Learning implementation for either segmentation or full-text extraction? The comments in the code suggest that, as of the last update, only the CRF Wapiti implementation can be used due to the large input sequence size.

- name: "segmentation"
  At this time, must always be CRF Wapiti; the input sequence size is too large for a Deep Learning implementation

- name: "fulltext"
  At this time, must always be CRF Wapiti; the input sequence size is too large for a Deep Learning implementation
